x/lockup/keeper/lock.go

```diff
@@ -780,3 +780,70 @@ func (k Keeper) getCoinsFromLocks(locks []types.PeriodLock) sdk.Coins {
 	}
 	return coins
 }
+
+// RebondTokens rebonds some amount or all coins of a lock with id lockID only if it is already unbonding
+func (k Keeper) RebondTokens(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coins sdk.Coins) error {
+	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return err
+	}
+
+	if lock.Owner != owner.String() {
+		return fmt.Errorf("lock %d is not owned by %s", lockID, owner)
+	}
+
+	if !lock.IsUnlocking() {
+		return fmt.Errorf("lock %d is not unlocking, rebonding only possible in unlocking stage", lockID)
+	}
+
+	// If all checks pass, we can rebond the tokens
+	var rebondedLock types.PeriodLock
+	// partial re-bonding
+	if coins != nil && !coins.IsEqual(lock.Coins) {
+		// This branch implies we will end up with two locks:
```

Comment:
basically, 3 out of 4 test cases test that branch, when `coinsToRebond` is `notFullRebondCoins` (which is set to 5), it will try to rebond these 5 tokens, out of initially locked 10 (`lockedCoins`)

Reply:
ahh that makes sense! I read the test wrong. Thanks!